pyccel/parser/semantic.py

```diff
@@ -2648,6 +2649,19 @@ def _visit_IfSection(self, expr, **settings):
 
     def _visit_If(self, expr, **settings):
         args = [self._visit(i, **settings) for i in expr.blocks]
+        allocations = [arg.get_attribute_nodes(Allocate) for arg in args]
+        var_shapes = [dict() for _ in range(len(args))]
+        for i,allocs in enumerate(allocations):
+            for a in allocs:
+                var_shapes[i][a.variable] = a.shape
+        variables = [v for branch in var_shapes for v in branch]
+        print(variables)
+
+        for v in variables:
+            if not all(v in branch_shapes.keys() for branch_shapes in var_shapes) \
+                    or not all(var_shapes[0][v]==branch_shapes[v] \
+                                for branch_shapes in var_shapes[1:]):
+                v.set_changeable_shape()
```

Comment:
I don't know if this optimization makes a noticeable difference, but `var_shapes[0][v]` could be stored in a temporary variable before the `if` statement.

Reply:
Fixed in ada1f0e1